src/server/pps/server/pipeline_controller.go

```diff
@@ -190,14 +190,17 @@ func (pc *pipelineController) step(timestamp time.Time) (isDelete bool, retErr e
 		return true, nil
 	}
 	var stepErr stepError
+	// TODO(msteffen) should this fail the pipeline? (currently getRC will restart
+	// the pipeline indefinitely)
+	rc, restart, err := pc.getRC(ctx, pi)
+	if restart != "" {
+		return false, pc.restartPipeline(ctx, pi, rc, restart)
+	}
+	if err != nil && !errors.Is(err, errRCNotFound) {
+		return false, err
+	}
 	errCount := 0
 	err = backoff.RetryNotify(func() error {
-		// TODO(msteffen) should this fail the pipeline? (currently getRC will restart
-		// the pipeline indefinitely)
-		rc, err := pc.getRC(ctx, pi)
-		if err != nil && !errors.Is(err, errRCNotFound) {
-			return err
-		}
```

Comment:
Hmm, I think I missed before, but I'm worried that when we moved refreshing pipeline info out of this retry loop we caused potential regressions on some of the reasons we adopted the top-level retry. Maybe worth thinking through this. Moving `getRC` out seems safer but I still worry that in the case of a retriable error we are going to run the logic on a version of the RC before any partial changes

Reply:
Totally ok if we decide not to commit to this part of the change. 

The reason I think it's worthwhile is that in the current state of things, functions like getRC have their own retry logic that becomes nested within the "top level" retry logic which I think makes things more difficult to reason about.

That being said, I didn't notice that potential error and agree that it's cause for concern, and maybe for moving all 3 operations into the top-level retry. It seems like the main class of retriable error is one where `setPipelineState()` fails, which I suppose we interpret as potentially being due to some DB race issue that prompt us to re-`step` the pipeline controller? If that's the case I think `Bump()` may be a little difficult to use for this - we wouldn't have a clear way to say "we retried enough times"

Reply:
that's a fair point. Yeah, a lot of what drove the earlier changes in this structure was avoiding nested retries, and I think that's a good goal. It just violates some assumptions in subtle ways - it probably doesn't matter as much here, as we don't have any transaction-type guarantees, but we'd be explicitly exposing ourselves to those issues rather than them mostly being limited to weird cases where users modify the cluster directly. I think on the balance this is good, and in the long term these decisions probably won't matter because more "pure" logic will naturally avoid these questions.